File: app/services/readiness.py
"""What the app still needs to know before it can search for you.

Two questions are answered here, and they are different. "Can a scan run at
all?" has exactly two requirements — what to search for and where — because
without them the only thing left to search with is somebody else's profile.
"Will the results be any good?" has many more, and none of them blocks: an
unknown fact never hides an offer (see ``candidate_facts``), it just stops
being able to rule one out.

The per-fact provenance already existed and nothing read it: ``matching-facts``
has answered with ``sources`` (cv / manuale / mancante) and a ``missing`` list
since 1.8.0, and the frontend rendered three rows of it. This is that idea
finished — one place that knows what is missing, so the profile page, the
dashboard checklist and the pre-scan gate cannot disagree.
"""

from __future__ import annotations

from dataclasses import asdict, dataclass
from typing import Any

from app.db import Database
from app.services.candidate_facts import candidate_facts
from app.services.search_intent import search_intent

BLOCKING = "blocking"
WARNING = "warning"

#: Which sub-tab of the profile page each item belongs to, so a panel can show
#: what is missing from ITS part instead of one undifferentiated list.
GROUP_TARGET = "target"
GROUP_ABOUT = "about"
GROUP_CONSTRAINTS = "constraints"

#: Facts read by ``candidate_facts``, and the tab that owns them.
_FACT_GROUP = {
    "years_experience": GROUP_CONSTRAINTS,
    "education_level": GROUP_CONSTRAINTS,
    "grade": GROUP_CONSTRAINTS,
    "degree_fields": GROUP_CONSTRAINTS,
    "driving_licence": GROUP_CONSTRAINTS,
    "protected_category": GROUP_CONSTRAINTS,
    "work_rule": GROUP_CONSTRAINTS,
}


@dataclass(frozen=True)
class ReadinessItem:
    """One thing the app knows, or does not."""

    id: str
    group: str
    severity: str
    #: ``ok`` | ``missing``. Deliberately not a boolean: "unknown" and "false"
    #: are different answers and this project has paid for confusing them.
    status: str
    #: Where the value came from when there is one: cv | manuale | dedotto.
    source: str = ""
    #: A short, already-readable value for the UI. Never a whole CV.
    value: str = ""


def _item(
    id_: str, group: str, severity: str, ok: bool, source: str = "", value: str = ""
) -> ReadinessItem:
    return ReadinessItem(
        id=id_,
        group=group,
        severity=severity,
        status="ok" if ok else "missing",
        source=source if ok else "",
        value=value if ok else "",
    )


def _has_watchlist(db: Database) -> bool:
    """Following employers is a way of saying what to look for."""
    if db.get_preference("watchlist_enabled", "0") not in ("1", "true", "on"):
        return False
    return bool(db.list_watchlist_companies(active_only=True))


def profile_readiness(db: Database) -> dict[str, Any]:
    intent = search_intent(db)
    facts = candidate_facts(db)
    profile = db.get_active_candidate_profile()
    watchlist = _has_watchlist(db)

    items: list[ReadinessItem] = [
        # The two that block. Everything else is advice.
        _item(
            "search_terms",
            GROUP_TARGET,
            BLOCKING,
            bool(intent.terms) or watchlist,
            source=intent.terms_origin if intent.terms else "watchlist",
            value=", ".join(intent.terms[:4]),
        ),
        _item(
            "location",
            GROUP_TARGET,
            BLOCKING,
            bool(intent.locations),
            source=intent.locations_origin,
            value=", ".join(intent.locations[:3]),
        ),
        _item(
            "cv",
            GROUP_ABOUT,
            WARNING,
            profile is not None,
            source="cv",
            value=str(profile.get("source_name") or "") if profile else "",
        ),
        _item(
            "goal",
            GROUP_TARGET,
            WARNING,
            bool((db.get_preference("onboarding_goal", "") or "").strip()),
            source="manuale",
            value=(db.get_preference("onboarding_goal", "") or "")[:80],
        ),
        _item(
            "ral_min",
            GROUP_TARGET,
            WARNING,
            bool((db.get_preference("onboarding_ral_min", "") or "").strip()),
            source="manuale",
            value=(db.get_preference("onboarding_ral_min", "") or ""),
        ),
    ]

    # The applicability facts, with the provenance candidate_facts already
    # tracks. A fact nobody stated is a warning: it cannot block an offer, so
    # it cannot block a scan either.
    readable = {
        "years_experience": "" if facts.years_experience is None else str(facts.years_experience),
        "education_level": facts.education_level or "",
        "grade": "" if facts.grade is None else str(facts.grade),
        "degree_fields": ", ".join(sorted(facts.degree_fields)),
        "driving_licence": "" if facts.driving_licence is None else str(facts.driving_licence),
        "protected_category": (
            "" if facts.protected_category is None else str(facts.protected_category)
        ),
        "work_rule": ", ".join(facts.work_rule.cities),
    }
    for name, group in _FACT_GROUP.items():
        origin = facts.sources.get(name, "mancante")
        items.append(
            _item(name, group, WARNING, origin != "mancante", source=origin, value=readable[name])
        )

    blocking = [i.id for i in items if i.severity == BLOCKING and i.status == "missing"]
    warnings = [i.id for i in items if i.severity == WARNING and i.status == "missing"]
    return {
        "ready": not blocking,
        "blocking": blocking,
        "warnings": warnings,
        "items": [asdict(i) for i in items],
        # The same chain the scan will use, so the search form can be filled in
        # with what the app would have used anyway — visibly, and editable.
        "suggested_terms": intent.terms,
        "terms_origin": intent.terms_origin,
        "suggested_locations": intent.locations,
        "locations_origin": intent.locations_origin,
    }

File: web/modules/readiness.js
// What the app still needs to know, shown where it can be answered.
//
// The server has always known which facts were missing — matching-facts has
// returned a `missing` list since 1.8.0 — and the page never read it. Now one
// answer feeds three places: the strip at the top of each profile tab, the
// checklist on the dashboard, and the gate in front of a scan. They cannot
// disagree, because there is only one of them.

import { api } from "./helpers.js";
import { t } from "./i18n.js";

let _cache = null;

//: Where each missing item can actually be answered. A strip that names a gap
//: without offering the field is just a complaint.
const FOCUS = {
  search_terms: { view: "job-search", el: "keywordsInput" },
  location: { view: "job-search", el: "locationsInput" },
  cv: { view: "profile", el: "cvFile" },
  goal: { view: "profile", el: "obGoal" },
  ral_min: { view: "profile", el: "obRalMin" },
  years_experience: { view: "profile", el: "factsEditBtn" },
  education_level: { view: "profile", el: "factsEditBtn" },
  grade: { view: "profile", el: "factsEditBtn" },
  degree_fields: { view: "profile", el: "factsEditBtn" },
  driving_licence: { view: "profile", el: "factsEditBtn" },
  protected_category: { view: "profile", el: "factsEditBtn" },
  work_rule: { view: "profile", el: "factCities" },
};

const SOURCE_KEY = {
  cv: "readiness.sourceCv",
  manuale: "readiness.sourceManual",
  last_scan: "readiness.sourceLastScan",
  shortlist: "readiness.sourceShortlist",
  profile: "readiness.sourceProfile",
  watchlist: "readiness.sourceWatchlist",
};

export async function fetchReadiness({ force = false } = {}) {
  if (_cache && !force) return _cache;
  try {
    _cache = await api("/api/profile/readiness");
  } catch {
    // A missing checklist must never be the reason something else fails.
    _cache = null;
  }
  return _cache;
}

export function invalidateReadiness() {
  _cache = null;
}

function _chip(item) {
  const label = t(`readiness.item.${item.id}`) || item.id;
  if (item.status === "ok") {
    const origin = SOURCE_KEY[item.source] ? t(SOURCE_KEY[item.source]) : "";
    const value = item.value ? `: ${item.value}` : "";
    return `<span class="readiness-chip is-ok" title="${origin}">${label}${value}</span>`;
  }
  const cls = item.severity === "blocking" ? "is-blocking" : "is-warning";
  return `<button type="button" class="readiness-chip ${cls}" data-readiness-fix="${item.id}">${label}</button>`;
}

/** Fill every `[data-readiness-group]` strip on the page. */
export async function renderReadinessStrips() {
  const strips = document.querySelectorAll("[data-readiness-group]");
  if (!strips.length) return;
  const report = await fetchReadiness({ force: true });
  if (!report) return;
  for (const strip of strips) {
    const group = strip.dataset.readinessGroup;
    const items = report.items.filter((i) => i.group === group);
    const missing = items.filter((i) => i.status === "missing");
    if (!missing.length) {
      strip.innerHTML = `<span class="readiness-done">${t("readiness.allSet")}</span>`;
      continue;
    }
    const blocking = missing.filter((i) => i.severity === "blocking");
    const optional = missing.filter((i) => i.severity !== "blocking");
    const section = (label, list) =>
      list.length
        ? `<div class="readiness-row"><span class="readiness-label">${label}</span>${list
            .map(_chip)
            .join("")}</div>`
        : "";
    strip.innerHTML =
      section(t("readiness.blocking"), blocking) + section(t("readiness.optional"), optional);
  }
}

/**
 * The pre-scan gate: two things, and only two.
 *
 * Everything else the app does not know is a warning, on the same principle
 * the scoring follows — an unknown fact never hides an offer, so it cannot
 * stop a scan either.
 */
export async function ensureProfileReady({
  showToast,
  revealElement,
  terms = [],
  locations = [],
  isRemote = false,
}) {
  // What is on the form right now answers the question just as well as what is
  // stored — the server has not seen these tags yet, and refusing a scan
  // someone has literally just typed the terms for would be absurd.
  const answered = new Set();
  if (terms.length) answered.add("search_terms");
  if (locations.length || isRemote) answered.add("location");
  if (answered.has("search_terms") && answered.has("location")) return true;

  const report = await fetchReadiness({ force: true });
  if (!report) return true; // never let a checklist be the reason a scan cannot run
  const blocking = report.blocking.filter((id) => !answered.has(id));
  if (!blocking.length) return true;
  const names = blocking.map((id) => t(`readiness.item.${id}`) || id).join(", ");
  showToast(t("readiness.scanBlockedBody", { items: names }), "error");
  const first = FOCUS[blocking[0]];
  if (first && revealElement) {
    revealElement(first.el);
    document.getElementById(first.el)?.focus?.();
  }
  return false;
}

/** Wire the "fill this in" chips once; they work on any strip. */
export function initReadiness({ revealElement }) {
  document.addEventListener("click", (event) => {
    const chip = event.target.closest("[data-readiness-fix]");
    if (!chip) return;
    const target = FOCUS[chip.dataset.readinessFix];
    if (!target) return;
    revealElement(target.el);
    document.getElementById(target.el)?.focus?.();
  });
}
